### PR TITLE
Upgrade to Jandex 3.2.0 and use the Jandex annotation overlay

### PR DESCRIPTION
**Pull request by Ladicek**
- https://smallrye.io/blog/jandex-3-2-0/

### CODE REVIEW

**Comment by Ladicek** [comment]
Added @mkouba and @manovotn to review the ArC commits and @geoand and @FroMage to review the RESTEasy Reactive commit.

**Review by geoand — APPROVED** [review]
Love it!

I only looked at the commit changing RESTEasy Reactive

**Review by manovotn — APPROVED** [review]
LGTM
Nice code cleanup!

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/AnnotationStore.java [hunk 1]
```diff
@@ -71,163 +50,33 @@ public Collection<AnnotationInstance> getAnnotations(AnnotationTarget target) {
...
-
-        @Override
-        public Transformation transform() {
-            if (LOG.isTraceEnabled()) {
-                String stack = Arrays.stream(Thread.currentThread().getStackTrace())
-                        .skip(2)
-                        .limit(7)
-                        .map(se -> "\n\t" + se.toString())
-                        .collect(Collectors.joining());
-                LOG.tracef("Transforming annotations of %s %s\n\t...", target, stack);
```
Comment (mkouba) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/AnnotationStore.java:165]: Does `AnnotationOverlay` from Jandex provide a similar trace logging? This code together with `AnnotationsTransformationContext.setAnnotations()` was very useful when debugging what transformer changed the final set of annotations...

**Review by mkouba — COMMENTED** [review]
I looked at non-BCE changes in ArC and it looks good! But it would be nice to have some numbers (in terms of memory consumption and build time) for a medium-sized application so that we can be sure there's no regression - after all, it is used extensively.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/AnnotationStore.java (hunk 1 above)
Comment (Ladicek) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/AnnotationStore.java:165]: It doesn't, no. Let me take another look at this.

Comment (Ladicek) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/AnnotationStore.java:165]: Added this logging back (to `Transformation`, which all current annotation transformations use). I'll think about adding similar logging to Jandex, but that's a bit more complex, as it doesn't have a concept of `Transformation` (and `Transformation.done()` in particular :-) ). All transformations are done directly on the `TransformationContext`.

**Comment by Sanne** [comment]
> I looked at non-BCE changes in ArC and it looks good! But it would be nice to have some numbers (in terms of memory consumption and build time) for a medium-sized application so that we can be sure there's no regression - after all, it is used extensively.

Agreed it would be great to verify metrics - but would you ask that as a blocker @mkouba ?  Just wondering about the status here, as I was eager to hit the merge button :)

**Comment by Ladicek** [comment]
I'm trying to get some numbers as we speak, actually :-) It's not easy, because a Quarkus build does a _lot_ of other things, so there's a lot of noise. I'll try to write a microbenchmark, but I fear it's just gonna end up stress testing the JIT compiler and not produce anything meaningful.

**Comment by mkouba** [comment]
> > I looked at non-BCE changes in ArC and it looks good! But it would be nice to have some numbers (in terms of memory consumption and build time) for a medium-sized application so that we can be sure there's no regression - after all, it is used extensively.
> 
> Agreed it would be great to verify metrics - but would you ask that as a blocker @mkouba ? Just wondering about the status here, as I was eager to hit the merge button :)

It's not a blocker but in my exprience, if we don't do it now, we never will ;-).

> I'm trying to get some numbers as we speak, actually :-) It's not easy, because a Quarkus build does a lot of other things, so there's a lot of noise. I'll try to write a microbenchmark, but I fear it's just gonna end up stress testing the JIT compiler and not produce anything meaningful.

I was thinking about manually testing something from https://github.com/quarkus-qe/beefy-scenarios and measure the build time/memory consumption. Just to verify there's no significant regression...

**Comment by Ladicek** [comment]
So I took https://github.com/Ladicek/arc-crazybeans/, which has 500 beans, and added the following extensions:

- quarkus-arc
- quarkus-cache
- quarkus-hibernate-orm
- quarkus-hibernate-validator
- quarkus-mailer
- quarkus-messaging
- quarkus-micrometer
- quarkus-rest
- quarkus-smallrye-fault-tolerance

The total number of ArC annotation transformations with these extensions, in the default configuration, is 19.

On an external computer (not on my work laptop), which executes very little processes and was configured for performance measurements, I repeatedly ran `mvn clean package` and looked at

```
[io.quarkus.deployment.QuarkusAugmentor] Quarkus augmentation completed in 1909ms
```

In current `main` branch, the reported time was nearly always between 1850 and 1900 ms, with several outliers between 1800 and 1850 ms and also between 1900 and 1950 ms.

With this PR, the reported time was nearly always between 1900 and 1950 ms, with several outliers between 1850 and 1900 ms and also between 1950 and 2000 ms.

To measure memory, I ran `MAVEN_OPTS="-XX:StartFlightRecording=filename=xxx.jfr,settings=profile,dumponexit=true" mvn clean package` 5 times for current `main` and 5 times for this PR. I then loaded all the JFR files in JMC and saw virtually no difference. The allocation patterns, the GC patterns, the amount of used memory, no significant change in any of those.

**Comment by mkouba** [comment]
> In current `main` branch, the reported time was nearly always between 1850 and 1900 ms, with several outliers between 1800 and 1850 ms and also between 1900 and 1950 ms.
> 
> With this PR, the reported time was nearly always between 1900 and 1950 ms, with several outliers between 1850 and 1900 ms and also between 1950 and 2000 ms.

Seems to be an acceptable drop. Did you try the jandex 3.2.0 or the main branch (including https://github.com/smallrye/jandex/commit/29acecceca1c3f75a98137f853cef23e70fe9109 ;-).

> To measure memory, I ran `MAVEN_OPTS="-XX:StartFlightRecording=filename=xxx.jfr,settings=profile,dumponexit=true" mvn clean package` 5 times for current `main` and 5 times for this PR. I then loaded all the JFR files in JMC and saw virtually no difference. The allocation patterns, the GC patterns, the amount of used memory, no significant change in any of those.

Cool.

**Review by mkouba — APPROVED** [review]

**Comment by Ladicek** [comment]
For memory, I only tried Jandex 3.2.0. For time, I tried Jandex 3.2.0 and Jandex `main` branch with an additional change (no `SENTINEL`, always store the set of annotations into the map), and I didn't see a difference.

**Pull request merged**